File: solutions/541.reverse-string-ii/reverse-string-ii.js
/**
 * @param {string} s
 * @param {number} k
 * @return {string}
 */
var reverseStr = function(s, k) {
    const chars = s.split('');
    let start = 0;
    while(start < s.length){
        let end = start + k - 1;
        if(end >= s.length - 1) end = s.length -1;
        swap(chars, start, end);
        start += 2*k;
    }
    return chars.join('');
    
    
    function swap(chars, start, end){
        while(end > start){
            const temp = chars[start];
            chars[start] = chars[end];
            chars[end] = temp;
            start++;
            end--;
        }
    }
    
};

File: solutions/541.reverse-string-ii/reverse-string-ii.java
class Solution {
    public String reverseStr(String s, int k) {
        int len = s.length();
        char[] chs = s.toCharArray();
        int start = 0;
        while(start < len){
            int end = start + k -1;
            if(end >= len -1) end = len -1;
            swap(chs, start, end);
            start += 2*k;
        }
        return String.valueOf(chs);
    }
    public void swap(char[] chs, int start, int end){
        while(end > start){
            char temp = chs[start];
            chs[start] = chs[end];
            chs[end] = temp;
            start++;
            end--;
        }
    }
}